File: backend/lib/craftup/game/recipe_level.ex
defmodule Craftup.Game.RecipeLevel do
  use Ecto.Schema
  import Ecto.Changeset

  schema "recipe_levels" do
    field :class_job_level, :integer
    field :difficulty, :integer
    field :durability, :integer
    field :quality, :integer
    field :stars, :integer
    field :suggested_control, :integer
    field :suggested_craftsmanship, :integer

    timestamps()
  end

  @doc false
  def changeset(recipe_level, attrs) do
    recipe_level
    |> cast(attrs, [
      :class_job_level,
      :stars,
      :suggested_craftsmanship,
      :suggested_control,
      :difficulty,
      :quality,
      :durability
    ])
    |> validate_required([
      :class_job_level,
      :stars,
      :suggested_craftsmanship,
      :suggested_control,
      :difficulty,
      :quality,
      :durability
    ])
  end
end

File: datasync/src/types/recipe_level.rs
use crate::types::GameItem;
use serde::{self, Deserialize};

#[derive(Debug, Deserialize)]
pub struct RecipeLevel {
    pub id: i64,
    pub class_job_level: i32,
    pub stars: i32,
    pub suggested_craftsmanship: i32,
    pub suggested_control: i32,
    pub difficulty: i32,
    pub quality: i32,
    pub durability: i32,
    pub unknown: i32,
}

impl GameItem for RecipeLevel {
    fn is_valid(&self) -> bool {
        self.class_job_level > 0
    }
}
